protos/file.proto

```diff
@@ -19,7 +19,191 @@
+//
+// Buffers:
+//  * There are no buffers
+// Children:
+//  * There are no children
+message ConstantEncoding {
+  // The constant value
+  // If this is empty then the value is null
+  bytes value = 1;
+}
+
+// Leaf encoding for fixed-width types
+//
+// This encoding does not store nulls, but it is commonly a
+// child of MaskedEncoding in which case some values may be
+// garbage.
+//
+// Buffers:
+//  * Values - A contiguous buffer of fixed-size values
+message ValueEncoding {
+  // The size of each item, in bits
+  uint32 item_width = 1;
+}
+
+// Densely encodes nullable types using a boolean mask
+//
+// Children:
+//  * Values - An array of values with the same logical length as this array
+//             Null values are present but the bytes are garbage and should be
+//             ignored.
+//  * Validity - A boolean array representing whether each value is null or not
+message MaskedEncoding {
+  // The encoding used to store the values
+  Encoding value_encoding = 1;
+  // The encoding used to store the validity
+  EncodingType validity_encoding = 2;
+}
+
+// Values encoded using a dictionary
+//
+// Buffers:
+//  * There are no buffers
+// Children:
+//  * An array of indices into the dictionary (which are unsigned integers)
+//  * An array of values
+message DictionaryEncoding {
+  // The number of bits in each key
+  uint32 key_width = 1;
+  // The encoding used to encode the keys
+  EncodingType key_encoding = 2;
+  // The encoding used to encode the values
+  EncodingType value_encoding = 3;
+}
+
+// Encoding for variable-length lists of data
+//
+// Buffers:
+//  * There are no buffers
+// Children:
+//  * An array of offsets (which are unsigned integer values)
+//  * An array of values
+message VariableLengthEncoding {
+  // The number of bits in each offset
+  uint32 offset_width = 1;
+  // If true, then lists with length 0 should be considered null values
+  bool empty_is_null = 2;
+  // The encoding used to encode the offsets
+  EncodingType offset_encoding = 3;
+  // The encoding used to encode the values
+  EncodingType value_encoding = 4;
+}
+
+// An encoding describes how an array is physically serialized into buffers
+//
+// Some encodings split an array into multiple arrays.  For example, plain
+// encoding stores an array of values and an array for validity.  Each of those
+// child arrays has its own encoding.  This means that the encoding forms a
+// tree of encodings where the leaves are buffers.
+message EncodingType {
+  oneof type {
+    ConstantEncoding constant = 1;
+    ValueEncoding value = 2;
+    MaskedEncoding masked = 3;
+    DictionaryEncoding dictionary = 4;
+    VariableLengthEncoding variable_length = 5;
+  }
+}
+
+message ColumnMetadata {
+  message Page {
+    // The file offsets of each of the page buffers
+    //
+    // These offsets might point to the column's data section or they
+    // might point to the column's metadata section.
+    //
```

Comment:
What does in the data vs metadata section?

Reply:
Data section is data that is unique to a page.  Metadata section is data that is common across all pages in a column.  E.g. if you wanted to have a single dictionary for all dictionary encoded pages in a column you could put it in the metadata section.